miniapp/miniapp_reduction_to_band.cpp

```diff
@@ -158,6 +160,11 @@ struct reductionToBandMiniapp {
         // wait and barrier for all ranks
         matrix.waitLocalTiles();
         taus.waitLocalTiles();
+        for (std::size_t i = 0; i < comm_grid.num_pipelines(); ++i) {
+          sync_wait(comm_grid.full_communicator_pipeline().exclusive());
+          sync_wait(comm_grid.row_communicator_pipeline().exclusive());
+          sync_wait(comm_grid.col_communicator_pipeline().exclusive());
+        }
```

Comment:
Waiting for communicators does not strictly seem to be required in all the miniapps, but I would add it to all of them in any case. Whether or not it's required probably depends on the internal communication patterns.

In any case, the above snippet seems to work fine in CI now. I think we could either:
- add a helper function available only for tests/miniapps with the signature `void wait_all_communicators(CommunicatorGrid&)` which does the above, or
- add a member function, similar to `Matrix::waitLocalTiles`, to `CommunicatorGrid` which does the above

Do you have preferences? Other ideas? I think the latter is nice in terms of encapsulation, but I'm not sure we want to encourage users to use it (yet...?).

Reply:
2nd option looks OK.

Reply:
I've added a helper `CommunicatorGrid::wait_all_communicators` in the latest commit. Unlike the snippet above which simply waits for the communicators (which does not ensure that communication has completed) I've made `wait_all_communicators` actually do an `MPI_Ibarrier`. At the same time I've removed the blocking call to `MPI_Barrier` at the end of the algorithm miniapps.

I have not changed _all_ `MPI_Barrier` calls to use `wait_all_communicators`. I think it makes sense, but if you agree to that change I will do that in a follow-up PR. What do you think?